File: C#/C00261172-Music-Application/C00261172-Music-Application/SelectableAudioList.cs
using C00261172_Music_Application.Application;
using C00261172_Music_Application.DataStructure;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace C00261172_Music_Application
{
    internal class SelectableAudioList
    {
        public Audio? GetSelectedAudio()
        {
            if (this.SelectedAudio == null)
            {
                this.SelectedAudio = this.AudioList.Head;
            }
            return this.SelectedAudio?.Data;
        }
        public Audio? GetNextMusic()
        {
            if (this.SelectedAudio == null)
            {
                this.SelectedAudio = this.AudioList.Head;
                return this.SelectedAudio?.Data;
            }
            this.SelectedAudio = this.SelectedAudio.Next;
            return this.SelectedAudio.Data;
        }
        public bool RemoveSelectedMusic()
        {
            return AudioList.RemoveData(GetSelectedAudio());
        }
        private DoubleLinkedNode<Audio>? SelectedAudio { get; set; }
        protected CircularLinkedList<Audio> AudioList = new();
    }
}

File: C++/C00261172-Music-Application/C00261172-Music-Application/SelectableAudioList.cpp
#include "SelectableAudioList.h"

std::shared_ptr<Audio> SelectableAudioList::getSelectedAudio()
{
	if (this->selectedAudio == nullptr) {
		this->selectedAudio = this->audioList.getHead();
		DoubleLinkedNode<Audio>* audioNode = this->selectedAudio.get();
		if (audioNode == nullptr) {
			return nullptr;
		}
		std::shared_ptr<Audio> audio = audioNode->data;
		if (audio.get() == nullptr) {
			return nullptr;
		}
		return audio;
	}
	return nullptr;
}

std::shared_ptr<Audio> SelectableAudioList::getNextSelectedAudio()
{
	if (this->selectedAudio == nullptr) {
		this->selectedAudio = this->audioList.getHead();
		return this->selectedAudio.get()->data;
	}
	this->selectedAudio = this->selectedAudio.get()->next;
	return this->selectedAudio.get()->data;
}

std::shared_ptr<Audio> SelectableAudioList::removeSelectedAudio()
{
	return std::shared_ptr<Audio>();
}
